### DISCUSSION TITLE
MDIForm example

### DISCUSSION THREAD

**Discussion by chunmingwang** [XusinboyBekchanov/VisualFBEditor#358] [Show and tell]
MDIForm example

[freebasic.net/forum](https://www.freebasic.net/forum/viewtopic.php?p=294158#p294158)

[MDIForm.zip](https://github.com/XusinboyBekchanov/VisualFBEditor/files/9412691/MDIForm.zip)

![Untitled](https://user-images.githubusercontent.com/35757455/186326539-ace57efc-34c2-40f7-ae74-0780c0cfb6a0.png)

**Comment by XusinboyBekchanov** [comment]
Very well.
In my opinion, you still need to do Delete to release resources when you close MDIChild.

**Comment by chunmingwang** [comment]
I will try to use Delele replace the `SendMessage(h, WM_CLOSE, 0, 0)` for improve

**Reply by hustbeef** [reply]
![image](https://user-images.githubusercontent.com/52040833/186364873-92d9f945-ccdd-49f1-86c4-b2ed1da35752.png)
Missing the ico in MDI form when the child form state is Maxmium size.

**Reply by XusinboyBekchanov** [reply]
But if you click on these places it works. In my opinion, the icon and buttons need to be drawn in dark mode.

**Reply by XusinboyBekchanov** [reply]
Fixed: Showing MDI Child Icon and Minimize, Restore, Close buttons:
https://github.com/XusinboyBekchanov/MyFbFramework/commit/81ff689e010abdfb698521ec468364db90c9235d

**Reply by XusinboyBekchanov** [reply]
On Dark mode

**Reply by chunmingwang** [reply]
thanks

**Comment by chunmingwang** [comment]
Updated using Delete release resources when close MDIChild

[MDIForm.zip](https://github.com/XusinboyBekchanov/VisualFBEditor/files/9415071/MDIForm.zip)

**Reply by XusinboyBekchanov** [reply]
When closing an empty MDIList, Tag gets 0 and it crashes.

**Reply by XusinboyBekchanov** [reply]
MDIChildType cannot delete itself. There will be a program crash.
```
Private Sub MDIChildType.Form_Destroy(ByRef Sender As Control)
	MDIForm.lstMdiChild.Remove(MDIForm.lstMdiChild.IndexOf(@This))
	If MDIForm.lstMdiChild.Count < 1 Then
		MDIForm.actMidChildIdx = -1
		MDIForm.StatusBar1.Text = ""
	End If
	'Delete @This
End Sub
```

**Reply by XusinboyBekchanov** [reply]
If you do this, there are no errors. We need to find other ways of delete.
```
Private Sub MDIFormType.mnuFileNew_Click(ByRef Sender As MenuItem)
	Static i As Integer = 0
	Dim frm As MDIChildType Ptr
	
	i += 1
	frm = New MDIChildType
	frm->Text = "Untitled - " & i
	frm->Show(MDIForm)
	frm->CloseForm
	Delete frm
End Sub
```

**Reply by XusinboyBekchanov** [reply]
Update Control.bas:
https://github.com/XusinboyBekchanov/MyFbFramework/commit/ff0c8291ecabbbcc8334afd25616fa445d0f0c39

After this change, your code works without a crash (except for MDIList):
```
Private Sub MDIChildType.Form_Destroy(ByRef Sender As Control)
	MDIForm.lstMdiChild.Remove(MDIForm.lstMdiChild.IndexOf(@This))
	If MDIForm.lstMdiChild.Count < 1 Then
		MDIForm.actMidChildIdx = -1
		MDIForm.StatusBar1.Text = ""
	End If
	Delete @This
End Sub
```

**Reply by chunmingwang** [reply]
thanks, i will check MDIList crashes issue later.

**Reply by chunmingwang** [reply]
Try this

```
Private Sub MDIListType.Form_Create(ByRef Sender As Control)
	Tag = -1
	If MDIMain.lstMdiChild.Count < 1 Then Exit Sub
	Dim i As Integer
	For i = 0 To MDIMain.lstMdiChild.Count - 1
		ListControl1.AddItem (Cast(MDIChildType Ptr, MDIMain.lstMdiChild.Item(i))->Text, MDIMain.lstMdiChild.Item(i))
	Next
	ListControl1.ItemIndex = MDIMain.actMidChildIdx
End Sub
```

**Reply by XusinboyBekchanov** [reply]
Yes, it will fix the program crash.

**Comment by hustbeef** [comment]
I think that the MDI child form is not in Darkmode. It will be showing if the MDI child form is in Darkmode.

**Reply by XusinboyBekchanov** [reply]
Here it is filled with dark color. Now we need to draw buttons and an icon:
![image](https://user-images.githubusercontent.com/32607344/186421983-3d4b681b-0f2d-436e-a491-8c79c6c1dc89.png)

**Comment by hustbeef** [comment]
I add code there:
` '#Region "Form"
	#include once "mff/Form.bi"
	#include once "mff/Menus.bi"
	#include once "mff/ReBar.bi"
	#include once "mff/ImageList.bi"
	#include once "mff/StatusBar.bi"
	#include once "mff/List.bi"

	Using My.Sys.Forms
	#ifdef __USE_WINAPI__
		InitDarkMode
		SetDarkMode(True, True)
	#endif
`
But the MDI child form is not in Darkmode.  How to set MDI child form is  in Darkmode. 
![image](https://user-images.githubusercontent.com/52040833/186427863-39affd49-2ccd-48f3-940a-ac1fbabbe52b.png)

**Comment by hustbeef** [comment]
OH. no need a MDIList.frm. we could add the Child windows to the menuItem of mennu "Windows".  The MDIList.frm is a sigle forms and the DarkMode work fine there.

**Comment by chunmingwang** [comment]
Update: Add the Child windows to the menuItem of mennu "Windows"

[MDIForm.zip](https://github.com/XusinboyBekchanov/VisualFBEditor/files/9422150/MDIForm.zip)

![Untitled](https://user-images.githubusercontent.com/35757455/186594995-52221898-b85f-472d-a187-cff9b0d1d2c6.png)

**Reply by XusinboyBekchanov** [reply]
Very good, you can send a request as an example.

**Reply by chunmingwang** [reply]
I create a request for MDIForm example
https://github.com/XusinboyBekchanov/VisualFBEditor/pull/359

**Comment by chunmingwang** [comment]
I create a request for MDIForm example
https://github.com/XusinboyBekchanov/VisualFBEditor/pull/359

**Reply by XusinboyBekchanov** [reply]
I accepted, thanks.

**Reply by chunmingwang** [reply]
closed the MDI child window will not raise the Destroy event, could you please help to check it.

**Reply by XusinboyBekchanov** [reply]
Answered in this comment:
https://github.com/XusinboyBekchanov/VisualFBEditor/issues/248#issuecomment-1236076472

**Reply by chunmingwang** [reply]
thanks
